### ISSUE TITLE
Support personal repositories not only GitHub organization

### ISSUE THREAD

**Issue by jugatsu** [terraform-providers/terraform-provider-github#45]
### Terraform Version
v0.9.5

### Affected Resource(s)
- github_repository

### Terraform Configuration Files
```hcl
provider "github" {
  token        = "${var.github_token}"
}

resource "github_repository" "otus-devops-terraform" {
  name        = "otus-devops-terraform"
  description = "Repository for public terraform code using in Otus DevOps course"
}

resource "github_repository" "practice-git-1" {
  name        = "practice-git-1"
  description = "Repository for practicing Git in Otus DevOps course"
}

resource "github_repository" "practice-git-2" {
  name        = "practice-git-2"
  has_issues  = true
  description = "Repository for practicing Git in Otus DevOps course"
}
```
### Debug Output
`terraform.tfstate`
```json
{
    "version": 3,
    "terraform_version": "0.9.5",
    "serial": 2,
    "lineage": "753781c3-ade7-4b8d-9723-2f7f9bbdb08b",
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {},
            "depends_on": []
        }
    ]
}
```

### Expected Behavior
It should create repository under personal account and write state in `terraform.tfstate`

### Actual Behavior
`terraform` asks `Github oraganization` when running with `terraform apply`. Also it returns `422` when running twice. So it isn't idempotent.

### Steps to Reproduce
`terraform apply`

**Comment by jugatsu** [comment]
https://github.com/jugatsu/otus-devops-terraform

**Comment by TraGicCode** [comment]
Any updates on this?

**Comment by paultyng** [comment]
Related to #77

**Comment by prdonahue** [comment]
I have a feeling this isn't going to happen for commercial reasons, but I'd love to be surprised.

I'm working on a blog post showing how to use Terraform along with GitHub and Hugo to create a static blog, but as soon as I tried `terraform apply` in my personal account ($7/mo) everything fails.

Can someone from GitHub either close this as "won't do" or let us know if this will be supported? If someone submits a PR enabling personal account will you accept it? cc @benbalter

**Comment by kquinsland** [comment]
the `username` is the `github_orginization` if you're operating under the personal/free account.

I was able to get `tf import github_repo...` to work with a personal access token.
I was able to run `tf apply` after importing and got a correct/successful "no changes needed" 
I then tried to change the name of the repo in terraform, and am now getting:
```github_repository.test: POST https://api.github.com/orgs/kquinsland/repos: 404 Not Found []```

Looking at the [API Docs](https://developer.github.com/v3/repos/#create) there is a different URL endpoint for creating in an org (`POST /orgs/:org/repos`) versus creating for a user (`POST /user/repos`.  I'm *very* new to `golang`, is there a significant amount of work involved to add a simple `account_type` value to the provider, where a value of `user` informs the provider to use the user version of the necessary APIs and a value of `org` uses the APIs that are currently there?

**Comment by p0bailey** [comment]
Having the possibilty to create and manage persona repos from TF would be amazing, especially for demos. Any updated down the line?

**Comment by p0bailey** [comment]
Hello there,
Meanwhile waiting for a fix into GitHub Terraform provider. I have just released a Terraform module for anyone interested to  set GitHup personal repositories. It's a bit raw, though.

https://github.com/p0bailey/terraform-github-personal-repositories

**Comment by binlab** [comment]
Any updates on this?

**Comment by p0bailey** [comment]
> Any updates on this?

Everything seems blocked.

**Comment by ahaffar** [comment]
i am using `github_provider` version 2.4.1 and it works for individual repos 
```
provider "github" {
  version = "> 2.4"
  token = var.token
  individual = false
  organization = var.organization
}
```
where organization is the `username`

**Comment by mrtazz** [comment]
this worked for me. With version `2.4.1` configured for my individual account as above I was at least able to:

- import state
- add a collaborator to a repo
- mark a repo as archived

**Comment by cornfeedhobo** [comment]
I am unable to confirm @mrtazz's success **_when importing_**:

```hcl
provider "github" {
    version    = "~> 2.4"
    individual = true
}

resource "github_repository" "test" {
    name = "go-travis-testing"
}
```

```bash
$ terraform init

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "github" (hashicorp/github) 2.4.1...
...snip...

$ terraform import github_repository.test cornfeedhobo:go-travis-testing
github_repository.test: Importing from ID "cornfeedhobo:go-travis-testing"...
github_repository.test: Import prepared!
  Prepared github_repository for import
github_repository.test: Refreshing state... [id=cornfeedhobo:go-travis-testing]

Error: This resource requires GitHub organization to be set on the provider.
```

@shoekstra @jcudit pinging you because I stumbled into this while attempting to test importing with `github_branch`.

**Comment by cornfeedhobo** [comment]
Hah. This is also broken when just applying normally.

```hcl
provider "github" {
    individual = true
}

resource "github_repository" "test" {
    name = "terraform-github_branch-test"
}
```

```bash
$ terraform apply
...snip...
github_repository.test: Creating...

Error: This resource requires GitHub organization to be set on the provider.

  on main.tf line 5, in resource "github_repository" "test":
   5: resource "github_repository" "test" {
```

**Comment by mrtazz** [comment]
@cornfeedhobo you need to set `individual = false` in the provider config. That's kinda the weird thing here. The provider then treats your individual account as an org, so there will definitely be edge cases that don't work.

**Comment by cornfeedhobo** [comment]
@mrtazz I tried that as well and it fails too.

```hcl
provider "github" {
    individual   = false
    organization = "cornfeedhobo"
}

resource "github_repository" "test" {
    name = "terraform-github_branch-test"
}
```

```plain
Error: GET https://api.github.com/orgs/cornfeedhobo: 404 Not Found []
```

Like others point out above, it's impossible because the API`s `/org` endpoint won't respond to usernames.

**Comment by mrtazz** [comment]
hmm that's curious. My provider config looks like this

```
provider "github" {
  version      = "> 2.4"
  individual   = false
  organization = "mrtazz"
}
```
And with that I was able to import all of my repos as well as set some things as mentioned in https://github.com/terraform-providers/terraform-provider-github/issues/45#issuecomment-603794145.

**Comment by cornfeedhobo** [comment]
Strange indeed! Every combination in the provider has failed for me

```plain
Error: This resource requires GitHub organization to be set on the provider.
Error: GET https://api.github.com/orgs/cornfeedhobo: 404 Not Found []
Error: If `individual` is true, `organization` cannot be set.
```

**Comment by cornfeedhobo** [comment]
@mrtazz are you exporting anything other than `GITHUB_TOKEN`?

**Comment by ozthegreat** [comment]
For **Personal** repos this works. It only works with 2.4.0 though, 2.4.1 and 2.5.0 break it.

```
# Configure the GitHub Provider.
# token from GITHUB_TOKEN env var (required by codepipline)
provider "github" {
  version    = "2.4.0"
  individual = false
  organization = "MyPersonalUserName"
}
```

**Comment by cornfeedhobo** [comment]
Interesting. As I sit here and finish a PR, I can already see how big of a fix this could end up being. I'll have to think about this a little. Maybe it would be best to fix a one or a few resources at a time. Limit the blast radius of each PR.

**Comment by jcudit** [comment]
To add some wider context, a `3.x` release of this provider was discussed over in https://github.com/terraform-providers/terraform-provider-github/pull/316 to solve a similar underlying problem for user identifiers.  If a `3.x` release helps to get this feature shipped, that can be arranged.

I initially thought [GraphQL](https://github.com/terraform-providers/terraform-provider-github/issues/83) support would be a cause for a major version bump as well.  However, there is a [plan](https://github.com/terraform-providers/terraform-provider-github/issues/83#issuecomment-600881472) to incrementally add support.  If we are to take a similar approach, comprehensive acceptance testing and state migration coverage will be needed to 🚢 this feature incrementally.

**Comment by mrtazz** [comment]
> @mrtazz are you exporting anything other than `GITHUB_TOKEN`?

@cornfeedhobo I don't have any other environment variables set.

@ozthegreat oh that's weird. I'm definitely running 2.4.1

```
Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "github" (hashicorp/github) 2.4.1...
```

**Comment by keyboardarmy** [comment]
this doesn't work in 2.6.0 either

**Comment by janaka** [comment]
Doesn't work for me with a personal repo either. Hopefully this information helps get this resolved.

I've tried 2.4 and 2.6. The issue with 2.4 is different to 2.6 it seem and might be caused by the GH API changing.

With provider version 2.4.0
```Error: GET https://api.github.com/repos/janaka/aws-app-mesh-example: 404 Not Found []```

With provider version 2.6.0
```Error: GET https://api.github.com/orgs/janaka: 404 Not Found []```

Here are the two relevant fragments of my Terraform

```  stage {
    name = "Source"

    action {
      name             = "SourceCode"
      category         = "Source"
      owner            = "ThirdParty"
      provider         = "GitHub"
      version          = "1"
      output_artifacts = ["SourceCodeArtifact"]

      configuration = {
        Owner  = "janaka"
        Repo   = "aws-app-mesh-example"
        Branch = "master"
      }
    }
}```

```provider "github" {
  version = "2.4.0"
  token        = local.webhook_secret
  individual = false
  organization = "janaka"
}```

```data "github_repository" "repo" {
  full_name = "janaka/aws-app-mesh-example"
}

resource "aws_codepipeline_webhook" "hello-world-api_webhook" {
  name            = "hello-world-api-webhook-github"
  authentication  = "GITHUB_HMAC"
  target_action   = "Source"
  target_pipeline = aws_codepipeline.codepipeline.name

  authentication_configuration {
    secret_token = local.webhook_secret
  }

  filter {
    json_path    = "$.commits[*]"
    match_equals = "hello-world-api/(src/.*|version)"
  }
}

# Wire the CodePipeline webhook into a GitHub repository.
resource "github_repository_webhook" "hello-world-api" {
  repository = data.github_repository.repo.name

  configuration {
    url          = aws_codepipeline_webhook.hello-world-api_webhook.url
    content_type = "json"
    insecure_ssl = false
    secret       = local.webhook_secret
  }

  events = ["push"]
}```

**Comment by chukaofili** [comment]
@janaka Did you check to make sure your token has the right permissions? you'll need repo admin when create you personal access token. This is using with provider version 2.4.0.

**Comment by janaka** [comment]
Pretty sure perms are good. I'll need to double check as it's been some days.

**Comment by janaka** [comment]
FYI - I moved the repo from my personal to an Org (the personal account owns) and all is now working. Same token with no perms changes. Provider 2.6.1.

**Comment by abitrolly** [comment]
@jcudit can you add this to https://github.com/terraform-providers/terraform-provider-github/milestone/8?closed=1 milestone.

**Comment by clebio** [comment]
Sorry, what is the fix here for the `github_repository` data source?  Nothing I see described above seems to work. I'm on `provider.github v2.9.2` and getting

> Error: GET https://api.github.com/orgs/clebio: 404 Not Found []

**Comment by cornfeedhobo** [comment]
@clebio pretty sure this was reverted

**Comment by janaka** [comment]
Given orgs became free around the same time, I just created one and moved my repo there.

**Comment by josjaf** [comment]
I'm now having the issue - @cornfeedhobo - how do you know it was reverted? I cannot easily create an organization because I'm part of a few already.

**Comment by cornfeedhobo** [comment]
@josjaf https://github.com/terraform-providers/terraform-provider-github/compare/v2.9.0...v2.9.1

**Comment by josjaf** [comment]
Any idea what the workaround is for personal repositories for now?

**Comment by cornfeedhobo** [comment]
@josjaf there is none that I know of right now. Sounds like most people have created orgs for the meantime.

**Comment by abitrolly** [comment]
@josjaf I moved to GitLab.

**Comment by ggsood** [comment]
This issue is now resolved now as per the provider version v4.1.0, you can create a github repo in your personal profile as 
```
provider "github" {
  owner = "<GITHUB_USERNAME>"
  token = "<GITHUB_TOKEN>"
}

resource "github_repository" "terraformed" {
  name        = "terraformed"
  description = "A repository created by Terraform"
  visibility  = "private"
} 
```
Thanks Community for fixing this issue.
